contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IAllowList.sol";
+import "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import "./INativeTokenDestination.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+// Precompiled Native Minter Contract Address
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// This is the address where the burned base fee go. We use this to send the amount of
+// burned tokens back to the source chain.
+address constant BURNED_TX_FEES_ADDRESS = 0x0100000000000000000000000000000000000000;
+// Designated Blackhole Address. This is where we burn token before sending an unlock
+// message to the source chain. Different from the burned tx fee address so we can
+// track each separately.
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000001;
+// Precompiled Warp address
+address constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
```

Comment:
Bumping TODO

Reply:
@minghinmatthewlam What's the best way to calculate an appropriate value for this?

Reply:
`forge` should be able to compute gas report, I believe `forge snapshot` or `forge test --gas-report`. Then I'd take a look at the gas usage for `receiveTeleporterMessage` on both source/destination. Mint gas requirements should match destination execution, and vice versa for transfer native gas requirements